### PR TITLE
[HIP] Fix issue w/ static initialization of function attributes

### PR DESCRIPTION
**Pull request by arghdos**
Fix issue w/ static initialization of function attributes on unspecialized type and add test.

This was a fun one to debug.  Basically, on line 214 (https://github.com/kokkos/kokkos/compare/develop...arghdos:fix_launch_mechs?expand=1#diff-078df6e850082d467d9c91f0888ce3b7b5f5f0fcde51bc91a00845ec6dd9bfb0R214), we use a static initialization of the function attributes to "cache" them.  However, this means that we will only get the function attributes for first kernel function pointer passed in.  This means that on subsequent calls we will get (e.g.,) incorrect `localSizeBytes`, which results in an incorrect launch bound determination at runtime.

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]
Looks reasonable to me.

File: core/src/HIP/Kokkos_HIP_KernelLaunch.hpp [hunk 1]
```diff
@@ -229,6 +231,9 @@ template <typename DriverType, unsigned int MaxThreadsPerBlock,
 struct HIPParallelLaunchKernelFunc<
     DriverType, Kokkos::LaunchBounds<MaxThreadsPerBlock, MinBlocksPerSM>,
     HIPLaunchMechanism::LocalMemory> {
+  using base_t = HIPParallelLaunchKernelFuncData<
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_KernelLaunch.hpp:234]: `base_t` is misleading, it is not the base class that `HIPParallelLaunchKernelFunc` derives from.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Looks good, except the name of the alias that is bad.  Please change it.

**Comment by dalg24** [comment]
By the way I'd swear we ran into that exact same kind of bug in the past

**Comment by dalg24** [comment]
> By the way I'd swear we ran into that exact same kind of bug in the past

Yup #3712

For the record this one was introduced in #3820

**Review by dalg24 — APPROVED** [review]

**Review by Rombur — APPROVED** [review]

**Pull request merged**